### DISCUSSION TITLE
socket.error: [Errno 115] Operation now in progress

### DISCUSSION THREAD

**Discussion by pokerfacett** [jtpereyda/boofuzz#569] [Q&A]
Traceback (most recent call last):
  File "boofuzz-http.py", line 75, in <module>
    main()
  File "boofuzz-http.py", line 71, in main
    session.fuzz()
  File "/usr/local/lib/python2.7/dist-packages/boofuzz/sessions.py", line 574, in fuzz
    self._main_fuzz_loop(self._iterate_protocol())
  File "/usr/local/lib/python2.7/dist-packages/boofuzz/sessions.py", line 677, in _main_fuzz_loop
    self._fuzz_current_case(*fuzz_args)
  File "/usr/local/lib/python2.7/dist-packages/boofuzz/sessions.py", line 1379, in _fuzz_current_case
    target.open()
  File "/usr/local/lib/python2.7/dist-packages/boofuzz/sessions.py", line 83, in open
    self._target_connection.open()
  File "/usr/local/lib/python2.7/dist-packages/boofuzz/socket_connection.py", line 158, in open
    self._sock.connect((self.host, self.port))
  File "/usr/lib/python2.7/socket.py", line 228, in meth
    return getattr(self._sock,name)(*args)
socket.error: [Errno 115] Operation now in progress

**Comment by pokerfacett** [comment]
@jtpereyda Help please!

**Comment by cq674350529** [comment]
I didn't come accross this error before. What target do you fuzz? Can you share your `boofuzz-http.py` or provide more details about your fuzzing context?

**Comment by jtpereyda** [comment]
One quick thing to try: install boofuzz from the latest source on master. I
recently PR’d some behavior that may help into master, but it hasn’t been
released yet.

**Comment by jtpereyda** [comment]
Installing from source: https://github.com/jtpereyda/boofuzz/blob/master/INSTALL.rst#from-source

**Comment by SR4ven** [comment]
> EINPROGRESS
The socket is nonblocking and the connection cannot be completed immediately. It is possible to select(2) or poll(2) for completion by selecting the socket for writing. After select(2) indicates writability, use getsockopt(2) to read the SO_ERROR option at level SOL_SOCKET to determine whether connect() completed successfully (SO_ERROR is zero) or unsuccessfully (SO_ERROR is one of the usual error codes listed here, explaining the reason for the failure).
https://linux.die.net/man/2/connect

This [answer](https://stackoverflow.com/a/10204791) to a similar problem guesses that this is caused by the `connect()` function taking longer than usual. Thus, all subsequent action on the socket will fail.

I can't remember seeing anything about nonblocking sockets in our socket_connection class, so I don't get why you are experiencing this issue. I will have a closer look at this later. [Ref](https://www.scottklement.com/rpg/socktut/nonblocking.html)

Could this be an issue caused by the target you are fuzzing being overloaded/stormed by the amount of packets you are sending? To prevent this you could set a `sleep_time` as a [session option](https://boofuzz.readthedocs.io/en/latest/source/Session.html)

Alternatively you could give the `reuse_target_connection` session option a try, as that doesn't reopen a new socket for every test case.
